examples/raytracing/example_raytracing.cpp

```diff
@@ -0,0 +1,300 @@
+
+  Kokkos::ScopeGuard guard(argc, argv);
+
+  namespace bpo = boost::program_options;
+
+  int num_spheres;
+  int num_rays;
+  float L;
+
+  bpo::options_description desc("Allowed options");
+  // clang-format off
+  desc.add_options()
+    ( "help", "help message" )
+    ("spheres", bpo::value<int>(&num_spheres)->default_value(100), "number of spheres")
+    ("rays", bpo::value<int>(&num_rays)->default_value(10000), "number of rays")
+    ("L", bpo::value<float>(&L)->default_value(100.0), "size of the domain")
+    ;
+  // clang-format on
+  bpo::variables_map vm;
+  bpo::store(bpo::command_line_parser(argc, argv).options(desc).run(), vm);
+  bpo::notify(vm);
+
+  if (vm.count("help") > 0)
+  {
+    std::cout << desc << '\n';
+    return 1;
+  }
+
+  std::cout << "ArborX version: " << ArborX::version() << std::endl;
+  std::cout << "ArborX hash   : " << ArborX::gitCommitHash() << std::endl;
+
+  std::uniform_real_distribution<float> uniform{0.0, 1.0};
+  std::default_random_engine gen;
+  auto rand_uniform = [&]() { return uniform(gen); };
+
+  // Random parameters for Gaussian distribution of radii
+  const float mu_R = 1.0;
+  const float sigma_R = mu_R / 3.0;
+
+  std::normal_distribution<> normal{mu_R, sigma_R};
+  auto rand_normal = [&]() { return std::max(normal(gen), 0.0); };
+
+  // Construct spheres
+  //
+  // The centers of spheres are uniformly sampling the domain. The radii of
+  // spheres have Gaussian (mu_R, sigma_R) sampling.
+  Kokkos::View<ArborX::Sphere *, MemorySpace> spheres(
+      Kokkos::ViewAllocateWithoutInitializing("spheres"), num_spheres);
+  auto spheres_host = Kokkos::create_mirror_view(spheres);
+  for (int i = 0; i < num_spheres; ++i)
+  {
+    spheres_host(i) = {
+        {rand_uniform() * L, rand_uniform() * L, rand_uniform() * L},
+        rand_normal()};
+  }
+  Kokkos::deep_copy(spheres, spheres_host);
+
+  // Construct rays
+  //
+  // The origins of rays are uniformly sampling the bottom surface of the
+  // domain. The direction vectors are uniformly sampling of a cosine-weighted
+  // hemisphere, It requires expressing the direction vector in the spherical
+  // coordinates as:
+  //    {sinpolar * cosazimuth, sinpolar * sinazimuth, cospolar}
+  // A detailed description can be found in the slides here (slide 47):
+  // https://cg.informatik.uni-freiburg.de/course_notes/graphics2_08_renderingEquation.pdf
+  Kokkos::View<ArborX::Ray *, MemorySpace> rays(
+      Kokkos::ViewAllocateWithoutInitializing("rays"), num_rays);
+  auto rays_host = Kokkos::create_mirror_view(rays);
+
+  for (int i = 0; i < num_rays; ++i)
+  {
+    float xi_1 = rand_uniform();
+    float xi_2 = rand_uniform();
+
+    rays_host(i) = {
+        ArborX::Point{rand_uniform() * L, rand_uniform() * L, 0.f},
+        ArborX::Ray::Vector{float(std::cos(2 * M_PI * xi_2) * std::sqrt(xi_1)),
+                            float(std::sin(2 * M_PI * xi_2) * std::sqrt(xi_1)),
+                            std::sqrt(1.f - xi_1)}};
+  }
```

Comment:
@wjge Now that I think about it, is there a reason to only shoot rays from the bottom of the domain? Why not do that from all the faces of the `[0, L]^3` box? It would give us a better opportunity to investigate sorting of rays.

Reply:
Not going to do this right now in this PR, can easily be added later.